libraries/terminal/src/Terminal.ts

```diff
@@ -0,0 +1,392 @@
+  }
+
+  /**
+   * {@inheritdoc ITerminal.registerProvider}
+   */
+  public registerProvider(provider: ITerminalProvider): void {
+    this._providers.add(provider);
+  }
+
+  /**
+   * {@inheritdoc ITerminal.unregisterProvider}
+   */
+  public unregisterProvider(provider: ITerminalProvider): void {
+    if (this._providers.has(provider)) {
+      this._providers.delete(provider);
+    }
+  }
+
+  /**
+   * {@inheritdoc ITerminal.write}
+   */
+  public write(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.log, false);
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeLine}
+   */
+  public writeLine(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.log, true);
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeWarning}
+   */
+  public writeWarning(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(
+      messageParts.map((part): string => Colorize.yellow(AnsiEscape.removeCodes(part))),
+      TerminalProviderSeverity.warning,
+      false
+    );
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeWarningLine}
+   */
+  public writeWarningLine(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(
+      messageParts.map((part): string => Colorize.yellow(AnsiEscape.removeCodes(part))),
+      TerminalProviderSeverity.warning,
+      true
+    );
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeError}
+   */
+  public writeError(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(
+      messageParts.map((part): string => Colorize.red(AnsiEscape.removeCodes(part))),
+      TerminalProviderSeverity.error,
+      false
+    );
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeErrorLine}
+   */
+  public writeErrorLine(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(
+      messageParts.map((part): string => Colorize.red(AnsiEscape.removeCodes(part))),
+      TerminalProviderSeverity.error,
+      true
+    );
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeVerbose}
+   */
+  public writeVerbose(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.verbose, false);
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeVerboseLine}
+   */
+  public writeVerboseLine(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.verbose, true);
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeDebug}
+   */
+  public writeDebug(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.debug, false);
+  }
+
+  /**
+   * {@inheritdoc ITerminal.writeDebugLine}
+   */
+  public writeDebugLine(...messageParts: string[]): void {
+    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.debug, true);
+  }
+
+  private _writeSegmentsToProviders(
+    segments: (string | ILegacyColorableSequence)[],
```

Comment:
Why is this considered "Legacy"? It's still a main part of the internal API. Is the plan to phase that out eventually?

Reply:
The API has been changed to only pass strings around. I renamed this to "Legacy" to maintain compatibility for things like plugins that were using the old `Colors` API that returned these instead of strings.